### ISSUE TITLE
ws.on :close does not give any event.reason for connection errors

### ISSUE THREAD

**Issue by SpComb** [faye/faye-websocket-ruby#102]
We are using the `ws = Faye::WebSocket::Client.new(...)` to connect to a websocket server, and logging the `ws.on :close do |event| info "connection closed with code #{event.code}: #{event.reason}" end`. However, it seems like this `event.reason` is always an empty string.

It would make debugging the websocket client configuration issues a ***lot*** easier if there were some indication of "connection timeout" vs `ECONNREFUSED` vs "you're connecting to a ws:// server using wss://`.

## Workaround
The following monkey-patch demonstrates passing through the `EventMachine::Connection#unbind` `Errno::*` as the close `event.reason`:

```ruby
class Faye::WebSocket::Client
  def emit_close(reason, code = 1006)
    @close_params = [reason, code]

    finalize_close
  end

  module Connection
    def unbind(reason = nil)
      if reason
        parent.__send__ :emit_close, reason
      else
        parent.__send__ :finalize_close
      end
    end
  end
end
```

Note that calling `begin_close` from the `unbind` is a bad idea, as the resulting `@stream.close_connection_after_writing` will not trigger a new `unbind`, and the client gets stuck.

#### Connecting to a server that is down:
```
kontena-agent | I, [2017-01-31T14:49:55.643239 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
kontena-agent | I, [2017-01-31T14:49:55.709221 #1]  INFO -- Kontena::WebsocketClient: connection closed with code 1006: Errno::ECONNREFUSED
```

#### Connecting to a `ws://` server using `wss://`
```
kontena-agent | I, [2017-01-31T14:53:14.353419 #1]  INFO -- Kontena::WebsocketClient: connecting to master at wss://192.168.66.1:9292
kontena-agent | I, [2017-01-31T14:53:15.212431 #1]  INFO -- Kontena::WebsocketClient: connection closed with code 1006: Errno::EPROTO
```

**Comment by jcoglan** [comment]
Thank you for bringing this to my attention, it is indeed a frustrating issue. Faye-websocket was designed for compatibility with the browser API, but browsers provide a bit of extra help here by logging out the cause of the error to the console, even if they provide no useful information via the events emitted.

I still think the `close` event should not carry this information, since the `code` and `reason` should to the extent possible reflect information conveyed via WebSocket close frames, not as a result of network errors. The correct thing to do here is emit an `error` event before `close`, and I think this would be an appropriate place to attach errors from the underlying I/O.

What do you make of this change? https://github.com/faye/faye-websocket-ruby/commit/39e26115b39c83e923dc0656548fad52dd736956

**Comment by SpComb** [comment]
Yes, this issue is entirely at odds with the design features of the browser API, which [explicitly](https://www.w3.org/TR/websockets/#concept-websocket-close-fail) forbids disclosing any useful debugging information :)

> User agents must not convey any failure information to scripts in a way that would allow a script to distinguish the following situations:
> * ...
>
> In all of these cases, the the WebSocket connection close code would be 1006, as required by the WebSocket Protocol specification. [WSP]
> Allowing a script to distinguish these cases would allow a script to probe the user's local network in preparation for an attack.
> In particular, this means the code 1015 is not used by the user agent (unless the server erroneously uses it in its close frame, of course).

So I suppose keeping the `close` event opaque to network connection transport errors makes some kind of bizarre sense, even though it would be the logical place to look, given the documented `reason` attribute.

> The correct thing to do here is emit an error event before close, and I think this would be an appropriate place to attach errors from the underlying I/O.

Extending the `error` event to contain useful information is probably cleaner. 

> What do you make of this change? 39e2611

Looks good.

Propagating it via `error` is also presumably going to be more consistent with any errors raised directly by the initial `EventMachine.connect`.

**Comment by SpComb** [comment]
Would it be possible to get a 0.10.x release with this fix?

**Comment by jcoglan** [comment]
Yes, sorry for the delay, it takes a while to run the pre-release testing for this package and I've had a bunch of distractions lately. I just released version 0.10.7.

**Comment by SpComb** [comment]
Thanks! That should make our websocket issues easier to debug now.
